feat: Auto set Party in Bank Transaction

marination opened the pull request:
Resolves part of https://github.com/frappe/erpnext/issues/34477
Ref: https://github.com/alyf-de/banking/issues/17
Docs: https://docs.erpnext.com/docs/v14/user/manual/en/accounts/bank-reconciliation

## Issue

- Bank Transaction holds almost no information about the Counter Party (the party we are paying/receiving from) as per the Bank Statement. 
- This counter party information is not likely to exactly match the Party in the system 
- Due to which the fields **Party Type** and **Party** are almost always blank. This makes identification hard in the **Bank Reconciliation Tool**
- Bank Transactions are usually bulk imported, hence setting these values manually are painful

## Enhancement

> **Note**: This matching is done fairly conservatively. It's better to not have a match than to have multiple wrong matches.

This PR adds an optional automatic Party matcher.

1. Enable it in **Accounts Settings**. Fuzzy matching can be optionally enabled
   <img width="1318" alt="Screenshot 2023-05-18 at 1 31 57 PM" src="https://github.com/frappe/erpnext/assets/25857446/9432397d-e5c9-4b41-a0a5-a67c113c8df5">

2. Register primary/most used bank details like _Account Number_ & _IBAN_ against **Customer**, **Supplier** & **Employee** via Bank Account or Employee DocType (for Employees). This is optional and will make the match more accurate if set by users.

3. Added fields to hold Counter party information in **Bank Transaction**.(_The population of these values will depend on integrations that import transactions or the users that upload the bank statement_ )
   <img width="1305" alt="Screenshot 2023-04-11 at 12 27 31 AM" src="https://user-images.githubusercontent.com/25857446/230973888-622d7380-d850-4d2f-b821-bdd5c09926dd.png">

4.  If Account details are absent in the Bank Transaction, **fuzzy matching** is _optionally_ done using the **Party Name (Bank Statement)** or **Description**. These are matched against Customer/Supplier/Employee names.
5. **The matching is done and values are set on Submit if auto matching is enabled**
6. Users can correct the Party & Party Type, if incorrect, after submission. This will only impact the edited transaction.
    <img width="1310" alt="Screenshot 2023-04-11 at 1 18 19 AM" src="https://user-images.githubusercontent.com/25857446/230985380-16644534-6524-4ac5-8ae0-652399e83e65.png">

## Working

<details><summary>Expand to view the working</summary>

![Automatch Flow](https://github.com/frappe/erpnext/assets/25857446/772744be-5b86-4ace-9a9d-415b05d58a46)

- **If more than one match was found against a description/party name, with the same score**, we skip matching and do not set a party in the bank transaction. We do not want to automate close calls, the user must decide these.

</details>

**TODO:**

- [x] Provision for user to correct mapping
- [x] Provision for users to create a mapping  (they can manually set the party in the transaction & also create a Bank Party Mapper)
- [x] Test
- [x] Docs
- [x]  Description
- [x] Use old bank fields to get party against bank data. Don't introduce new ways of storing bank data.
- [x] Optional fuzzy matching + If multiple similar matches, skip automatching

> **For the reviewer:** Hit me up for test data

marination committed:
feat: Store Party bank details in party records (Customer/Supplier/Employee/Shareholder)

marination committed:
feat: Party auto-matcher from Bank Transaction data

- Created Bank Party Mapper
- Created class to auto match by account/iban or party name/description(fuzzy)
- Automatch and set in transaction or create mapper
- `rapidfuzz` introduced

marination committed:
chore: Single query with or filter to search Party Mapper by name/desc

marination committed:
fix: Don't set description as key in Mapper doc if matched by description

- Description is volatile and will keep changing
- It will lead to multiple Bank Party Mapper docs for the same party that will never be referenced again
- Parts of the descripton keep changing which is why it will never match a mapper record
- If matched by desc, dont create mapper record.

marination committed:
feat: Manually Update/Correct Party in Bank Transaction

- On updating bank trans.n party after submit, the corresponding mapper doc will be updated too
- The mapper doc in turn will update all linked bank transactions that do not have this updated value
- Added Bank Party Mapper hidden link in Bank Transaction
- Rename field in BPM to `Party Name` as it does not hold description data
- If a BT matches with a BPM record, link that record in the BT

marination committed:
chore: Perform automatch on submit

- misc: Clearer naming

marination committed:
chore: Make auto matching party configurable

- Checkbox in Accounts settings "Enable Automatic Party Matching"
- Check before invoking automatching methods
- misc: Remove TODO comments

marination committed:
fix: Match by both Account No and IBAN & other cleanups

- A BT could have both account and iban, and a Supplier could have only IBAN set
- In this case, matching by either (only account) gives no match
- Match by Account OR IBAN, use `or_filters`
- If matched, set both account no. and IBAN in Bank Party Mapper

- Explain AutoMatchParty
- Add type hints to return values
- Use `set_value` to set values in BT after matching since its an after submit event

marination committed:
Merge branch 'develop' into bank-trans-party-automatch

marination committed:
Merge branch 'develop' into bank-trans-party-automatch

marination committed:
test: Match by Account No, IBAN, Party Name, Desc and match correction

marination committed:
fix: Remove bank details fields from Shareholder

ankush commented on pyproject.toml:
```diff
@@ -12,6 +12,7 @@ dependencies = [
     "pycountry~=20.7.3",
     "Unidecode~=1.2.0",
     "barcodenumber~=0.5.0",
+    "rapidfuzz~=2.15.0",
```
_Just commenting on dependency_

This has compiled components but wheels seem to be available for most py versions on linux, also seems to be actively maintained so it should be fine.

marination commented:
@ Whoever reviews this: Hit me up for test data

barredterra reviewed (commented):
Tested the matching, LGTM 👍🏼 

Two doubts:

- [x] Why add bank details to **Customer** and **Supplier** instead of fetching linked **Bank Accounts**?
- [x] Why **Bank Party Mapper** instead of searching old **Bank Transactions**, which should hold the same information?
   -> "If the match was found on fuzzy matching 'description', no mapper record is created". We could no make this distinction when searching old **Bank Transactions**.

marination commented:
> Why add bank details to Customer and Supplier instead of fetching linked Bank Accounts?

@deepeshgarg007 Could use your inputs here.

There's no uniform way to register Bank Accounts with Parties 😢.
For Employee: Bank details are registered in separate fields in the doctype
For other parties: We have to select a Party in the Bank Account. But if there are 2 parties that share the same account then we have to duplicate the Bank Account and link the party

What would be ideal is multiple parties linked to a Bank Account, this way everything is in the same place. 
What I did here was add Bank Account No. and IBAN fields to three party types. But that's perhaps creating yet another way to link bank data to party. It's all ...messy 😵‍💫

Any suggestions ?

marination committed:
Merge branch 'develop' into bank-trans-party-automatch

marination committed:
Merge branch 'develop' into bank-trans-party-automatch

marination commented:
@deepeshgarg007 any thoughts ?

vorasmit commented:
How about working with matching rules (like we have tax rules or pricing rules)?

Users create rules based on the amount or description or reference no or any bank transaction field for that matter.
eg: If the description contains "ABC" and the Deposit amount is positive --> Party Type is Customer and Party is ABC Ltd.
or If the description contains "Interest" and the Deposit amount is positive --> Suggest the creation of Interest Entry (with Interest account)

Can be extended to cover any other use cases that may be useful.

marination commented:
@vorasmit  Thanks for your inputs 😺
> If the description contains "ABC" and the Deposit amount is positive --> Party Type is Customer and Party is ABC Ltd

This kind of already happens. Party matches are given priority depending on the if it is a withdrawal or deposit. If it is a deposit then we are receiving a payment, that means Customer should be matched first. Employee and Supplier are matched if nothing is found among Customers, because it is possible (but not often) that a Supplier gives you a refund which is a deposit in your account. 
`If the description contains "ABC" ` is checked via fuzzy matching, creating rules for this seems kind of redundant here and way too much work for a user since bank transaction descriptions are very volatile.
Users can manually create a Bank Party Mapper entry linking the account number/IBAN/account holder name to the Party which will correctly match in the future. Again description is left out due to volatility at the moment.

Also matching rules sort of suggest that there are too many possibilities/ways to match the party that they need to be configured. But I see very limited things for a quick match at the moment that I have tried my best to incorporate. 

**What could change perhaps is** giving users the option to select which field must be matched? Depending on what is used to fetch bank transactions, some fields could be useless. Like the Plaid integration only works with the description field, so the Party Name/ Account Holder becomes kinda useless at the moment. That can be determined in the future, it's a rather small feature/fix.

> If the description contains "Interest" and the Deposit amount is positive --> Suggest the creation of Interest Entry (with Interest account)

I believe this falls out of scope of this PR. We're just trying to match a party here to make it easier in the Bank Reconciliation Tool. But this does fall in the scope of https://github.com/frappe/erpnext/issues/34477 as a whole, feel free to leave a comment on the issue or collaborate, there's a lot more we have to do to close the issue :)

vorasmit commented:
I appreciate your efforts here.

Problems would not arise if there is an exact match with the account number and the party is set.

However, there are certain limitations of fuzzy matches.

Where there are similar party names, it would be either always correct or always incorrect. This in turn adds responsibility to check additionally if the party set is correct every time.

This could be beneficial in the long run, if there was additional logic built-in, to look for historical matches giving it a priority.

How are descriptions usually set by the bank and how rules could make things more firm?
With respect to India, if NEFT or RTGS is performed, the description would contain the name of the party as added to net banking. If a cheque is issued or received, it would contain the name as per the cheque.

Where a rule is created, it's almost certain of what the user is expecting. Hence the recommendation. The second example, I gave earlier was just to showcase how it could be extended to simplify reconciliation. This is how it is done in Zoho Books, and it seems to be useful.

barredterra commented:
> Where there are similar party names, it would be either always correct or always incorrect. This in turn adds responsibility to check additionally if the party set is correct every time.

@vorasmit the way Marica implemented this, you'll only have to fix wrong matches once. The system remembers and in the future this party/account will always be matched correctly.

vorasmit commented:
For the fuzzy match, I am not sure if a mapper is created. It's not possible to create the same against the description either.

https://github.com/frappe/erpnext/pull/34675/files#diff-9b46ef4b0625a3eeb63b99694b64928323b206ed346ac92d95d5788299549acbR25

Edit:
For a client, we did the recent implementation, they have over 20k customers. Many are similar and need suffixes as to from which area they belong. 

Final thoughts:  If you want many people to use this feature, it is important to ensure that it does not generate unexpected results.

marination commented:
You have a valid concern here. Many similar names could cause chaos.
I think your rules suggestion can be incorporated into bank party mapper. I'll keep the scope to party matching for now and work with the use case you provided. 
Knowing which key to look for is helpful, since we don't have that right now we resort to fuzzy matching. Thanks @vorasmit

marination committed:
fix: Use existing bank fields to match by bank account no/IBAN

- Remove newly added fields in Party doctypes to store bank details
- Use Bank Account's fields to match against account no/iban
- For employee, if Bank Account does not exist, find in Employee doctype against account no/iban

marination committed:
fix: Tests

marination committed:
feat: Optional Fuzzy Matching & Skip Matches for multiple similar matches

- Fuzzy matching can be enabled optionally in the settings
- If a query gets multiple matches with the same score, do not set a party as it is an extremely close call
- misc: Add 'cancelled' status to Bank transaction
- Test for skipping matching with extremely close matches

marination committed:
Merge branch 'develop' into bank-trans-party-automatch

marination commented:
@deepeshgarg007 Docs and description has been updated. The latest changes resolve the following ambiguities:

> Why add bank details to Customer and Supplier instead of fetching linked Bank Accounts?
   
Bank details are searched for in Bank Account now. In the case of **Employee**, the Bank Account is searched and then the Employee doctype as it has bank account fields in it historically.
No new ways to store bank details

> @vorasmit's concern: "For a client, we did the recent implementation, they have over 20k customers. Many are similar and need suffixes as to from which area they belong."

I did brainstorm this quite a bit and discussed this with @barredterra. Smit suggested rules but that would cause users to create way too many rules to match their plethora of parties. Could not really come up with anything else that wasn't bad UX.

So **fuzzy matching is optional**. In the docs, users are encouraged to test the feature out first. The matching might be a wee bit problematic where extremely similar names exist (e.g. Adithya Medical& Surgical vs Adithya Medical & Surgical vs Adithya Medical And Surgical) but will work great for data where there is not so much similarity.

I also increased the match score cutoff to 80.

Additionally fuzzy matching does not set any value if multiple matches with the same score were obtained for a query.
E.g. if the query in the Bank Transaction is "A.K. Medical" and two parties exist such as "A.K. Medical Hall" and "A.K. Medical Store", both will be accurate matches. In this case we do not make the system decide between close calls and leave it blank for the user to set it themselves or not set it at all.

barredterra committed:
Merge branch 'develop' into bank-trans-party-automatch

marination committed:
chore: Remove Bank Party Mapper implementation

- Matching by Acc No/IBAN can easily happen with Bank Accounts. It's not a tedious query
- Historical lookups for  Party Name/Desc match are very tricky. The user could have manually set a match and we would not know. Also this leaves the Bank Party Mapper only useful for Party Name/Desc lookups, which feels excessive.
- We want to reduce the number of places the same data is stored and reduce confusion
- The Party Name/Desc will optionally happen fuzzily, or not at all
- There will be no Mapper lookups

marination committed:
Merge branch 'develop' into bank-trans-party-automatch

marination committed:
chore: Remove instances of `bank_party_mapper` and use `new_doc`

barredterra committed:
Merge branch 'develop' into bank-trans-party-automatch

barredterra reviewed (approved):
Tested matching by party name, description and IBAN. LGTM.

Todo: Update docs and PR description (to not confuse future readers).

marination commented:
@barredterra docs and PR description updated ✅

barredterra committed:
Merge branch 'develop' into bank-trans-party-automatch

marination committed:
Merge branch 'develop' into bank-trans-party-automatch
